src/text_shaper/directwrite_shaper.cpp

```diff
@@ -66,6 +66,60 @@ namespace
         }
         return font_slant::normal;
     }
+
+    void renderGlyphRunToBitmap(IDWriteGlyphRunAnalysis* _glyphAnalysis, RECT& _textureBounds, bitmap_format _targetFormat, std::vector<uint8_t>::iterator& _it)
+    {
+        const auto width = _textureBounds.right - _textureBounds.left;
+        const auto height = _textureBounds.bottom - _textureBounds.top;
+
+        std::vector<uint8_t> tmp;
+        tmp.resize(height *width * 3);
+
+        auto hr = _glyphAnalysis->CreateAlphaTexture(
+            DWRITE_TEXTURE_CLEARTYPE_3x1,
+            &_textureBounds,
+            tmp.data(),
+            tmp.size()
+        );
+
+        for (auto i = 0; i < height; i++)
+            for (auto j = 0; j < width; j++)
+            {
+                const auto base = ((height - 1 - i) *width + j) * 3;
+                const auto r = tmp[base];
+                const auto g = tmp[base + 1];
+                const auto b = tmp[base + 2];
+                if (r > 0)
```

Comment:
This is my naive implementation to composite multiple layers of bitmap.

Reply:
Oh this is certainly an interesting part of code that i have too sirens some more time in trying to understand ASAP. :)

Reply:
I don't honestly know if this is the correct way to do so 😅  I don't have much graphic programming experience.

Reply:
OK this is not the correct way to composite the different layers. This is the correct way: https://github.com/qt/qtbase/blob/615a9cf99150cfee34c14cfed0aa9583f5993934/src/gui/text/windows/qwindowsfontenginedirectwrite.cpp#L849

But I don't really understand the process. Also I still can't get it working correctly...